### DISCUSSION TITLE
How to get token_type_ids and attention_mask and get the result text?

### DISCUSSION THREAD

**Discussion by ljzj2** [microsoft/onnxruntime#32572] [Mobile Q&A]
I am using android application 10
I have get InferenceSession and tokenizer but I searched that I have to get token_type_ids and attention_mask but there is `EncodingToIds` 
there is no token_type_ids.

what I want
`1.ask questions about my txt file.`

```
 var vocabPath = Path.Combine(CacheDir.AbsolutePath, "vocab.txt");
 using (var s = Assets.Open("Models/vocab.txt"));
     using (var fs = File.Create(vocabPath));
         s.CopyTo(fs);

         fs.Dispose();

         tokenizer = BertTokenizer.Create(vocabPath, new BertOptions { RemoveNonSpacingMarks = true });

         var encodings = tokenizer.EncodeToIds("This is a sample");

         var path = Path.Combine(CacheDir.AbsolutePath, "allmini.onnx");
         using (var s1 = Assets.Open("Models/model.onnx"))
         {
             using (var fs1 = File.Create(path))
             {
                 s1.CopyTo(fs1);
             }
         }
         _session = new InferenceSession(path);
var inputs = new[]
{
    NamedOnnxValue.CreateFromTensor("input_ids",idTensor),
};

var results = _session.Run(inputs);
```

there is the repo`https://github.com/ljzj2/XMedicalAndroid`

**Comment by identity-wael** [comment]
There are two separate issues here: constructing the model inputs and choosing a model that can produce the answer you want.

In the linked repository's tensor-building loop, `idTensor[0, i] = 1` puts **1 in every occupied input-ID position**. Those positions need the actual tokenizer IDs. The ones belong in the attention mask instead:

```csharp
// idTensor and maskTensor are separate tensors with matching [1, length] shapes.
for (int i = 0; i < encodings.Count; i++)
{
    idTensor[0, i] = encodings[i];
    maskTensor[0, i] = 1;
}
```

Ensure `encodings.Count` fits the allocated length and the model's limit. Fill unused input positions with the tokenizer's padding ID and leave their mask entries at zero. Keep required special tokens in the encoded sequence. For a single BERT sequence, segment IDs (`token_type_ids`) are normally all zero; paired sequences use different segments.

Before adding inputs, inspect `_session.InputMetadata` for the **actual required names, element types, and dimensions**. Only supply `token_type_ids` if this ONNX export expects it; use `long` tensors only if it expects int64. Also inspect `_session.OutputMetadata` instead of assuming the output is text.

If `allmini.onnx` is an export of **all-MiniLM-L6-v2**, it is an embedding model: its [model card](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2) describes 384-dimensional sentence vectors for tasks such as semantic search. These floats cannot be decoded into an answer using the tokenizer. Some exports include pooling; others return token embeddings that need attention-mask-aware pooling.

For questions about a text file, you can use MiniLM to embed chunks of the file and retrieve chunks similar to the question. A separate generative model can then answer using those chunks, or your app can simply display the retrieved passages. Could you share the model's original download/export link and the input/output metadata? That would establish the exact inputs and whether pooling is already included.

**Comment by ljzj2** [comment]
thank you.

> fill unused input positions with the tokenizer's padding ID and leave their mask entries at zero

"this is a sample" to input_ids only have 4 I have to add other 0 to the input_ids?

yes `allmini.onnx` is an export of `all-MiniLM-L6-V2`  you said these floats cannot be decoded into an answer using tokenizer so I can not get an answer?

`you can use MiniLM to embed chunks of the file and retrieve chunks similar to the question` how to do it?

I am sorry the model's original download link is in other repos and  I believe it is download in huggingface. and input/output metadata where can I copy it? I have a vocab.txt.


thank you again .

I wonder if there is any chance give me an example? thank you.
